File: 2585.java
import java.io.*;
public class Main
{
 static int[][] map=new int[10][10];
 static int[][]	p=new int[4][4];
 public static void main(String[] args)throws IOException
 {
  InputStreamReader is=new InputStreamReader(System.in);
  BufferedReader in=new BufferedReader(is);
  while(true)
  {
   String s=in.readLine();
   if(s.equals("ENDOFINPUT"))break;
   String[] ss;
   for(int i=0;i< 4;i++)
   {
	ss=in.readLine().split(" ");
	for(int j=0;j< 4;j++)
		p[i][j]=Integer.parseInt(ss[j]);
   }
   in.readLine();
   for(int i=1;i< 10;i++)
     for(int j=1;j< 10;j++)
         map[i][j]=0;
   f();
   for(int i=1;i< 10;i++)
	map[i][i]=0;
   while(true)
   {
      boolean qq=false;
      for(int i=1;i< 10;i++)
	for(int j=1;j< 10;j++)
          for(int k=1;k< 10;k++)
	    if(map[i][j]==1&&map[j][k]==1&&map[i][k]==0)
	    {
		map[i][k]=1;
		qq=true;
	    }
      if(!qq)break;
    }
    boolean bb=true;
    for(int i=1;i< 10;i++)
	if(map[i][i]==1){
          bb=false;
	  break;
	}
    if(bb) System.out.println("THESE WINDOWS ARE CLEAN");
    else System.out.println("THESE WINDOWS ARE BROKEN");
   }
 }

  static void f()
  {
        int x;
 	x=p[0][1];
	map[x][1]=1;
	map[x][2]=1;
	x=p[0][2];
	map[x][2]=1;
	map[x][3]=1;
	x=p[1][0];
	map[x][1]=1;
	map[x][4]=1;
	x=p[2][0];
	map[x][4]=1;
	map[x][7]=1;
	x=p[1][3];
	map[x][3]=1;
	map[x][6]=1;
	x=p[2][3];
	map[x][6]=1;
	map[x][9]=1;
	x=p[3][1];
	map[x][7]=1;
	map[x][8]=1;
	x=p[3][2];
	map[x][8]=1;
	map[x][9]=1;
	x=p[1][1];
	map[x][1]=1;
	map[x][2]=1;
	map[x][4]=1;
	map[x][5]=1;
	x=p[1][2];
	map[x][2]=1;
	map[x][3]=1;
	map[x][5]=1;
	map[x][6]=1;
	x=p[2][1];
	map[x][4]=1;
	map[x][5]=1;
	map[x][7]=1;
	map[x][8]=1;
	x=p[2][2];
	map[x][5]=1;
	map[x][6]=1;
	map[x][8]=1;
	map[x][9]=1;
   }
}

File: 2585.cc
#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include<string>
#include <vector>
#define INF 1E9
using namespace std;
vector<int> map[16];
bool ok[10][10];
int vis[10];
void init()
{
    int i,now=0;
    for(i=1;i<10&&now<15;i++)
    {
        map[now].push_back(i);
        map[now+1].push_back(i);
        map[now+4].push_back(i);
        map[now+5].push_back(i);
        now++;
        if(i%3==0)now++;
    }
    return;
}
bool dfs(int m)
{
    int i;
    vis[m]=-1;
    for(i=1;i<10;i++)
    {
        if(!ok[m][i])continue;
        if(vis[i]<0)return 0;
        if(!vis[i]&&!dfs(i))return 0;
    }
    vis[m]=1;
    return 1;
}
bool topo()
{
    for(int i=1;i<10;i++)
      if(!vis[i]&&!dfs(i))return 0;
    return 1;
}
int main()
{
    init();
    string s;
    int i,t,j;
    while(cin>>s&&s!="ENDOFINPUT")
    {
        memset(vis,0,sizeof(vis));
        memset(ok,0,sizeof(ok));
        for(i=0;i<16;i++)
        {
           scanf("%d",&t);
           for(j=0;j<map[i].size();j++)
           {
               if(t==map[i][j])continue;
               ok[t][map[i][j]]=1;
           }
        }
        if(topo())cout<<"THESE WINDOWS ARE CLEAN"<<endl;
        else cout<<"THESE WINDOWS ARE BROKEN"<<endl;
        cin>>s;
    }
}


